### PR TITLE
All-in-one Makefile

### PR DESCRIPTION
**Pull request by duglin**
First pass at a single Makefile to build everything.

I think this is easier to maintain and it supports doing everything
using Docker by just defining the DOCKER env var - eg:
	DOCKER=1 make ...

Right now there are some targets missing (like pushing) but I wanted
to get a sense of how people felt about this direction before I worked
on the last bits.

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

**Comment by pmorie** [comment]
Movement in the right direction, LGTM

File: Makefile [hunk 1]
```diff
@@ -1,85 +1,92 @@
...
-$(SUB): %.sub:
-	$(ECHO) for dir in $(DIRS); do $(MAKE) --no-print-directory -C "$${dir}" $* || exit $$? ; done
+# Util targets 
+##############
 
-init:
-	$(ECHO) glide install --strip-vendor
+verify: init
+	@echo Running verify:
+	@for i in contrib pkg; do \
```
Comment (MHBauer) [inline Makefile]: `cmd/` ?

**Comment by MHBauer** [comment]
LGTM

File: DEVGUIDE.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Developer's Guide to Service-Catalog
```
Comment (martinmaly) [inline DEVGUIDE.md]: Should https://github.com/kubernetes-incubator/service-catalog/blob/master/docs/BUILD_SYSTEM.md be deleted?

File: DEVGUIDE.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+To build the service-catalog you have two options:
+* `make build`
+* `DOCKER=1 make build`
+
+Both will build all of the executables, into the `bin` directory. However,
+the second option will do the build within a Docker container - meaning you
+do not need to have all of the necessary tooling installed on your host
+(such as a golang compiler or glide). Whichever option you choose, the
+results should be the same.
```
Comment (martinmaly) [inline DEVGUIDE.md]: I worry that on a mac, DOCKER=1 will build Linux binaries (which is fine) but will overwrite any binaries built without DOCKER=1 (which doesn't seem right).

File: Makefile [hunk 4]
```diff
@@ -1,85 +1,92 @@
...
-lint: lint.sub
-push: push.sub
-test: test.sub
-
-clean: clean.sub
-	rm -rf $(BINDIR)
-	rm -f .dockerInit
-	rm -f .scBuildImage
-	docker rmi -f scbuildimage > /dev/null 2>&1 || true
+all: build test verify
```
Comment (martinmaly) [inline Makefile]: should` hack/*.mk` be deleted?

File: Makefile [hunk 5]
```diff
@@ -1,85 +1,92 @@
...
+  -v $(PWD):/go/src/github.com/kubernetes-incubator/service-catalog \
+  scbuildimage
+endif
+
+ROOT=$(dir $(abspath $(lastword $(MAKEFILE_LIST))))
+BINDIR=$(ROOT)/bin
+export GOPATH=$(shell cd ../../../..;pwd):$(PWD)/vendor
+COVERAGE?=$(CURDIR)/coverage.html
+SC_PKG=github.com/kubernetes-incubator/service-catalog
+SRC_FILES=$(shell sh -c "find contrib pkg -name \\*.go | sort | uniq")
```
Comment (martinmaly) [inline Makefile]: unused

File: Makefile [hunk 6]
```diff
@@ -1,85 +1,92 @@
...
+	$(DOCKER) hack/coverage.sh --html "$(COVERAGE)" $(addprefix ./,$(TEST_DIRS))
 
-.PHONY: apiserver
-apiserver:
-	go install -v github.com/kubernetes-incubator/service-catalog/cmd/service-catalog
-	go build -v -o $(BINDIR)/apiserver cmd/service-catalog/server.go
+test: 
+	@echo Running tests:
+	@for i in $(SRC_PKGS); do \
+	  $(DOCKER) go test $$i ; \
```
Comment (martinmaly) [inline Makefile]: My experience with for loops is that they won't fail if one of the iterations fails (see line 71 of previous version of the Makefile).

File: Makefile [hunk 7]
```diff
@@ -1,85 +1,92 @@
...
+	done
+
+clean:
+	rm -rf $(BINDIR)
```
Comment (martinmaly) [inline Makefile]: what principle do you apply when deciding whether to use `@` or not? (compare lines 70 and 89, and 84 for example)

File: Makefile [hunk 8]
```diff
@@ -1,85 +1,92 @@
...
-
-ALL := all build build-linux build-darwin clean docker push test lint coverage
-SUB := $(addsuffix .sub, $(ALL))
-.PHONY: $(ALL) $(SUB) format
-
-# Recursive targets.
-build: build.sub
-build-linux: build-linux.sub
-build-darwin: build-darwin.sub
-docker: docker.sub
```
Comment (martinmaly) [inline Makefile]: are we removing ability to produce Docker images? That will break e2e tests we run on Jenkins.

File: DEVGUIDE.md (hunk 3 above)
Comment (duglin) [inline DEVGUIDE.md]: I didn't touch on the multi-arch build stuff yet - wanted to wait until we're sure we want to head this way.
How does it work today? What are the default exe? the local arch/os or linux?

File: Makefile (hunk 8 above)
Comment (duglin) [inline Makefile]: Not removing it, just didn't get there yet.

File: Makefile (hunk 4 above)
Comment (duglin) [inline Makefile]: if we adopt this path, yes

File: Makefile (hunk 1 above)
Comment (duglin) [inline Makefile]: fixed

File: Makefile (hunk 5 above)
Comment (duglin) [inline Makefile]: removed

File: Makefile (hunk 6 above)
Comment (duglin) [inline Makefile]: I was wondering about that - thanks!

**Comment by arschles** [comment]
Does this close https://github.com/kubernetes-incubator/service-catalog/issues/73?

File: Makefile (hunk 4 above)
Comment (arschles) [inline Makefile]: @duglin can you put a TODO list in the PR description? something like this:

TODO:

- [ ] remove `hack/*.mk`
- [ ] remove/migrate bits from `BUILD_SYSTEM.md` according to new build logic
- [ ] ...

File: Makefile [hunk 9]
```diff
@@ -1,85 +1,100 @@
...
+	@echo Running verify:
+	@for i in $(TOP_DIRS); do \
+      $(DOCKER) gofmt -l $$i || exit $$? ; \
+	done > .out 2>&1
+	@bash -c '[ "`cat .out`" == "" ] || \
+	  (echo -e "\n*** Please 'gofmt' the following:" ; cat .out ; echo ; false)'
+	@rm .out
+	@for i in $(SRC_PKGS); do \
+	  ($(DOCKER) golint --set_exit_status $$i/... && \
+	  $(DOCKER) go vet $$i )|| exit $$? ; \
```
Comment (arschles) [inline Makefile]: I don't suppose there's any black `Makefile` magic to run these (and other applicable) steps in parallel?

**Comment by martinmaly** [comment]
I think when done it should close #73 
One thing I've been wondering is ... how close is this new approach to _the_ canonical k8s approach? Is it a copy of what k8s has, or a custom implementation? If custom ... should we just copy everything k8s has re: build scripts and avoid having to do it ourselves?

**Comment by arschles** [comment]
LGTM, assuming rebase goes smoothly. None of my comments are blocking. Ping when rebase is complete and I'll merge

File: Makefile (hunk 9 above)
Comment (arschles) [inline Makefile]: Yea, I figured. Thanks for answering

**Comment by duglin** [comment]
rebased

**Pull request merged**